deprecate `SimpleExecutor` (#18753)
bevyengine/bevy: release-content/migration-guides/simple_executor_going_away.md

---
title: Removed Simple Executor
pull_requests: [18753]
---

Bevy has deprecated `SimpleExecutor`, one of the `SystemExecutor`s in Bevy alongside `SingleThreadedExecutor` and `MultiThreadedExecutor` (which aren't going anywhere any time soon).
The `MultiThreadedExecutor` is great at large schedules and async heavy work, and the `SingleThreadedExecutor` is good at smaller schedules or schedules that have fewer parallelizable systems.

> So what was `SimpleExecutor` good at? Not much. That's why we plan to remove it. Removing it will reduce some maintenance and consistency burdons on maintainers, allowing them to focus on more exciting features!

Reviewer: ```suggestion
So what was `SimpleExecutor` good at? Not much. That's why we plan to remove it. Removing it will reduce some maintenance and consistency burden, allowing us to focus on more exciting features!
```

If you were using `SimpleExecutor`, consider upgrading to `SingleThreadedExecutor` instead, or try `MultiThreadedExecutor` if if fits the schedule.

> It's worth mentioning that `SimpleExecutor` runs deferred commands inbetween *each* system, regardless of it it was needed.

Reviewer: This needs to come earlier, and talk more about the UX side of things.

Author: Good call. Clearly justifying this to users is important. It's a lot better now I think. Thanks!

The other executors are more efficient about this, but that means they need extra information about when to run those commands.
In most schedules, that information comes from the contents and ordering of systems, via `before`, `after`, `chain`, etc.
If a schedule that was previously using `SimpleExecutor` still needs commands from one system to be applied before another system runs,
make sure that ordering is enforced explicitly by these methods, rather than implicitly by the order of `add_systems`.

> If you face unresolvable issues migrating from `SimpleExecutor`, please make an issue on github!

Reviewer: ```suggestion
If you have a strong preference for the existing behavior of `SimpleExecutor`, please make an issue and we can discuss your needs.
```

The `SimpleExecutor` leaves performance on the table compared to the other executors in favor of simplicity.
Specifically, `SimpleExecutor` applies any commands a system produces right after it finishes, so every system starts with a clean `World` with no pending commands.
As a result, the default `SimpleExecutor` runs all systems in the order they are added to the schedule, though more ordering constraints can be applied, like `before`, `after`, `chain`, etc.
In other executors, these ordering onstraints also inform the executor exactly where to apply commands.
For example, if system `A` produces commands and runs `before` system `B`, `A`'s commands will be applied before `B` starts.
However, the `before` ordering is implicit in `SimpleExecutor` if `A` is added to the schedule before `B`.

The dueling behavior between ordering systems based on when they were added to a schedule as opposed to using ordering constraints is difficult to maintain and can be confusing, especially for new users.
But, if you have a strong preference for the existing behavior of `SimpleExecutor`, please make an issue and we can discuss your needs.

> If you were using `SimpleExecutor`, consider upgrading to `SingleThreadedExecutor` instead, or try `MultiThreadedExecutor` if if fits the schedule.

Reviewer 2: ```suggestion
If you were using `SimpleExecutor`, consider upgrading to `SingleThreadedExecutor` instead, or try `MultiThreadedExecutor` if it fits the schedule.
```
Small typo :)

Author: Nice catch!
